lsbatch/man1/bjdep.1
.ds ]W %
.ds ]L
.nh
.TH post 1 "OpenLava Version 3.3 - Mar 2016"
.br
.SH NAME
\fBbjdep\fR - displays job dependencies
.SH SYNOPSIS
.BR
.PP
.PP
\fBbjdep\fR \fIjob_ID\fR | \fB"\fR\fIjob_ID\fR\fB[\fR\fIindex\fR\fB]"
.PP
\fBbjdep\fR\fB \fR[\fB-h\fR | \fB-V\fR]
.SH DESCRIPTION
.BR
.PP
.PP
Allows user to display all job dependencies. It outputs a list of other jobs 
that a job depends on.
.SH OPTIONS
.BR
.PP
.TP
 \fIjob_ID\fR | \fB"\fR\fIjob_ID\fR\fB[\fR\fIindex\fR\fB]"
\fR
.IP
Required. Job ID of the job or job array on which to operate.
.IP
For an array job, the index, the square brackets, and the quotation
marks are required.
.IP
Displays all jobs that this job depends on.

.SH EXAMPLE
.BR
.PP
.TP
$bjdep 135918
.PP
Dependencies of job 135918
.br
DEPENDENCY  JOBID  JOB_STATUS   DEPENDENCY_STATUS
.br
done        135917   PEND         false
.br
exit        135816   PEND         false
.br
done        135815   RUN          false

.SH SEE ALSO
.BR
.PP
.PP
bsub(1), bjobs(1)


NEWS
Release notes for OpenLava versions

3.3.pre.2
=========
-- Keep the lsb.events and lsb.acct files always open to improve
   file access performance.
-- Implement the switching of the lsb.events asynchronously in mbatchd
   child process.
-- Implemented the bjdep command. bjdep jobid prints the status of
   job dependencies for the given job.
-- Add the bjdep man page.
-- Implement NIOS_RWAIT_SELECT, if configure in lsf.conf then nios
   and ls_rwait() will not use the SIGUSR1 to detect a task that
   has finished. ls_rwait() will select() on a connection with nios
   instead and nios will send a message when a task exits.
-- Fix RES NIOS protocol bug that prevented multiple tasks running on
   the same host from being collected using ls_rwait().
-- Implemented job groups. New commands are bgadd, bgdel and bgroups.
-- Implemented bjobs -UF unformatted bjobs output.
-- Add LSF_BINDIR in openlava startup scripts.
-- Implement bjobs -g to display only selected job groups.
-- Implement bkill -g to kill only jobs in a specific job group.

3.3.pre.1
=========
-- Fix fairshare algorithm to decrease the number of jobs per user
   only if the job has been started.
-- Fix a bug in ownership policy where wrong runtime comparison was
   performed.
-- Implement the replay of job absolute runtime for ownership policy.
-- Implement floating client feature. If the user configures in lsf.conf
   LIM_ACCEPT_FLOAT_CLIENT=y any host can submit jobs and query the
   daemons without being registered in the lsf.cluster file.

3.2.1
======
-- Fix the parsing algorithm for compact host name.

3.2.0
======
-- Fix main Makefile.am to include the m4 directory.

3.2.beta1
=========
-- Hosts member od a host group can now be specified in a compact format
   for example: (ol[1-3]).
-- Implement configure option --enable-debug. If set the compilation will
   have the -Werror option enabled. By default this option is not set.

3.2.pre
========
-- badmin ckconfig/reconfig displays configuration error and warning
   messages.
-- bsub/bmod/brestart -h displays usage information.
-- Added USER_SHARES information in lsb.users.5
-- Add lsclusters command.
-- Use qsort() to sort jobs in MBD this improves the performance
   of replay and general bsub operations.
-- Introduce a getc2() functions to read lsb.events in memory
   before parsing it.
-- Reduce the number of malloc operations when reading the
   lsb.events file.
-- Implement memory reservation if the queue has SLOT_MEMORY_RESERVE configured
   and the job requests rusage of memory. The jobs has to reserve slots
   on the hosts as well.
-- Implement a correct switching of the lim.events file.
-- Update the bmgroup command to display GROUP_SLOT and MAX_SLOTS
   sections if configured.
-- Update the bqueues command to display slot and memory reservation.
-- Allow batch hosts configured in lsb.hosts to be specified in a
   compact form, i.e. ol[1-3] the mbatchd will expand example this into 3
   hosts ol1, ol2, ol3.
-- Allow OpenLava to use cores as cpus regardless is the machines has
   hyperthreads or not. To enable this behaviour set the variable
   LIM_DEFINE_NCPUS=cores|threads in lsf.conf. Default is threads.
-- Fix bsub bug in which if user specify -n min,max the max was ignored.
-- Make fairshare configuration compatible with the LSF one.

3.1.4
======

-- Fix the problem that if the order of events in lim.events is not correct
   LIM may core dump.
-- Change the ownership of the mbatchd log file to avoid messages going
   to syslog messages file.
-- Prevent LIM from core dumping in findHost() if the input name is NULL.
-- Enable the secondary admin to run 'badmin reconfig'.
-- Fix res nios protocol to pass the correct taskid to nios.

3.1.3
======
-- Corrected the lsf.cluster man page describing LIM communication
   parameters.
-- Enable LIM communication debugging using LC_COMM class and
   LOG_INFO level.
-- Fix bug in announceMaster() which was slowing down announcement
   of slave hosts.
-- Fix a bug in the slave host timer causing slave to send load
   to master more frequently then necessary.
-- Fix counter update in fairshare policy when jobs are suspended
   and resumed.
-- Remove spurious res log messages.

3.1.2
=====
-- Fix a critical mbatchd bug when sorting queues.
-- Enable the openlava.spec file and the rpm.sh script
   which builds the rpm.

3.1.1
=====
-- Fix scheduling bug in which performance could drop if the lowest
   priority queue was always full.
-- Fix the incorrect encoding of bchkpnt command data structure.

3.1.0
======
-- Modify openlava.sh to forcefully terminate sbatchd if SIGTERM does
   not work.
-- Fix LIM bug in which the code was accessing non initialized memory.
-- Fix several memory leaks in LIM.
-- Fix res hang in child handler.
-- Fix memory corruption in mbatchd when dealing with rusage.
-- Fix mbatchd child core dump when processing bugroup command.
-- Improve fairshare performance.
-- Introduce an lsf.conf parameter MBD_MAX_JOB_SCHED which allows site
   to configure how many jobs should MBD attempt to schedule per cycle.
   Typically two times the number of slots.
-- Introduce an lsf.conf parameter MBD_QSORT_JOBS which changes the way
   jobs are inserted and sorted in the PJL. This parameter avoids searching
   the job list upon a new job arrival speeding up the submission and overall
   MBD operations.
-- Fix issues in the job sorting order while using qsort().
-- Increase the array size of LIM clients from 64 to 2014.
-- Fix lim core dump if the lim receives an operation request from
   outside of the OL cluster and it is not LIM_ADD_HOST.
-- Fix bsub < file which was not creating the shell file correctly.
-- Fix rusage bug in which the host could be allocated to a job even with
   not enough resources.
-- Fix memory error in mbatchd when computing job's rusage.
-- Fix various memory leaks

3.1pre1
=======

-- Removed obsolete mls code and infrastructure simplifying user
   management.

lsbatch/cmd/bjdep.c
/*
 * Copyright (C) 2016 David Bigagli
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */
#include "cmd.h"
#include <string.h>
#include <ctype.h>

static void
usage(void)
{
    fprintf(stderr, "bjdep: [ -h ] [ -V ] jobId\n");
}

static void
print_jobdep(int, struct job_dep *, LS_LONG_INT);

int
main(int argc, char** argv)
{
    int cc;
    LS_LONG_INT jobid;
    struct job_dep *jobdep;

    if (lsb_init(argv[0]) < 0) {
	lsb_perror("lsb_init");
	return -1;
    }

    while ((cc = getopt(argc, argv, "hV")) != EOF) {
	switch (cc) {
	    case 'V':
		fputs(_LS_VERSION_, stderr);
		return 0;
	    case 'h':
		usage();
		exit(-1);
	}
    }

    if (argc <= optind) {
	usage();
	return -1;
    }

    if (getOneJobId(argv[argc - 1], &jobid, 0)) {
	usage();
	return -1;
    }

    jobdep = lsb_jobdep(jobid, &cc);
    if (jobdep == NULL) {
	if (lsberrno == LSBE_NO_ERROR) {
	    fprintf(stderr, "\
bjdep: No job dependencies for job %s.\n", lsb_jobid2str(jobid));
	    return -1;
	}
	fprintf(stderr, "bjdep: %s.\n", lsb_sysmsg());
	return -1;
    }
    print_jobdep(cc, jobdep, jobid);

    free_jobdep(cc, jobdep);

    return 0;
}

/* print_jobdep()
 */
static void
print_jobdep(int cc, struct job_dep *jobdep, LS_LONG_INT jobid)
{
    int i;

    printf("Dependencies of job %s\n", lsb_jobid2str(jobid));

    printf("%-10s  %-5s  %-10s   %-17s\n",
	   "DEPENDENCY", "JOBID", "JOB_STATUS", "DEPENDENCY_STATUS");

    for (i = 0; i < cc; i++) {
	printf("\
%-10s  %-5s   ", jobdep[i].dependency, jobdep[i].jobid);
	if (IS_PEND(jobdep[i].jstatus))
	    printf("%-10s   ", "PEND");
	else if (IS_START(jobdep[i].jstatus))
	    printf("%-10s   ", "RUN");
	else if (jobdep[i].jstatus & JOB_STAT_DONE)
	    printf("%-10s   ", "DONE");
	else if (jobdep[i].jstatus & JOB_STAT_EXIT)
	    printf("%-10s   ", "EXIT");
	else
	    printf("%-10s   ", "BOH?");
	if (jobdep[i].depstatus == true)
	    printf("%-s", "true");
	else
	    printf("%-s", "false");
	printf("\n");
    }
}



